contracts/metagov/utils/VoteEscrowTokenManager.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "../../refs/CoreRef.sol";
+import "../../core/TribeRoles.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+interface IVeToken {
+    function balanceOf(address) external view returns (uint256);
+    function locked(address) external view returns (uint256);
+    function create_lock(uint256 value, uint256 unlock_time) external;
+    function increase_amount(uint256 value) external;
+    function increase_unlock_time(uint256 unlock_time) external;
+    function withdraw() external;
+    function locked__end(address) external view returns (uint256);
+    function checkpoint() external;
+}
+
+/// @title Vote-escrowed Token Manager
+/// Used to permanently lock tokens in a vote-escrow contract, and refresh
+/// the lock duration as needed.
+/// @author Fei Protocol
+abstract contract VoteEscrowTokenManager is CoreRef {
+
+    /// @notice The maximum lock duration of veTokens
+    uint256 private immutable MAXTIME;
+    /// @notice One week, in seconds. Vote-locking is rounded down to weeks.
+    uint256 private constant WEEK = 7 * 86400; // 1 week, in seconds
+
+    /// @notice The vote-escrowed token address
+    IVeToken public immutable veToken;
+
+    /// @notice The token address
+    IERC20 public immutable liquidToken;
+
+    /// @notice VoteEscrowTokenManager token Snapshot Delegator PCV Deposit constructor
+    /// @param _liquidToken the token to lock for vote-escrow
+    /// @param _veToken the vote-escrowed token used in governance
+    /// @param _maxTime maximum time tokens can be vote-escrowed for
+    constructor(
+        IERC20 _liquidToken,
+        IVeToken _veToken,
+        uint256 _maxTime
+    ) {
+        liquidToken = _liquidToken;
+        veToken = _veToken;
+        MAXTIME = _maxTime;
+    }
+
+    /// @notice Deposit tokens to get veTokens. Set lock duration to MAXTIME.
+    /// The only way to withdraw tokens will be to pause this contract
+    /// for MAXTIME and then call exitLock().
+    function lock() external whenNotPaused onlyTribeRole(TribeRoles.METAGOVERNANCE_TOKEN_STAKING) {
+        uint256 tokenBalance = liquidToken.balanceOf(address(this));
+        uint256 locked = veToken.locked(address(this));
+        uint256 lockHorizon = ((block.timestamp + MAXTIME) / WEEK) * WEEK;
```

Comment:
the lock horizon is rounded down to entire weeks in the veToken contract, I am doing the same here to make sure I can test that the lock horizon is > what is already locked, because otherwise the call to increase_lock_time reverts (so I don't want to call it, if it's to set the same lock horizon)